biolink2/example.rkt

```diff
@@ -54,32 +54,51 @@
                  (mat 'close)
                  (printf "Finished processing ~s rows\n" count)))))
 
-;; materialize the (concept curie property value) relation if not present
-(unless (directory-exists? (db-path "concept"))
-  (printf "buildling relation: ~s; ~s\n"
-          (db-path fnin.nodeprop) (db-path "concept"))
-  (let/files ((in (db-path fnin.nodeprop))) ()
-    (materialize-dsv-stream
-      in header.nodeprop
-      ;; input columns
-      '(curie property value) buffer-size (db-path "concept")
-      ;; attribute names, types, and internal key name for indexing
-      '(curie property value) '(string string string) 'id
-      ;; primary table columns (without any sorted-columns)
-      '(((curie property value) . ())
-        ;; index table columns (without any sorted-columns)
-        ((value property id)    . ())))))
+;; materialize a relation if not present
+(define (materialize-relation name fnin header fields types)
+  (unless (directory-exists? (db-path name))
+    (printf "buildling relation: ~s; ~s\n"
+            (db-path fnin) (db-path name))
+    (let/files ((in (db-path fnin))) ()
+      (materialize-dsv-stream
+        in header
+        ;; input columns
+        fields buffer-size (db-path name)
+        ;; attribute names, types, and internal key name for indexing
+        fields types 'id
+        ;; primary table columns (without any sorted-columns)
+        `((,fields . ())
+          ;; index table columns (without any sorted-columns)
+          (,(append (cdr fields) (list 'id)) . ()))))))
+
+(materialize-relation "concept" fnin.nodeprop header.nodeprop
+                      '(curie property value)
+                      '(string string string))
+
+(materialize-relation "predicate" fnin.edge header.edge
+                      '(:id :start :end)
+                      '(string string string))
+
+(materialize-relation "predicateprop" fnin.edgeprop header.edgeprop
+                      '(:id property value)
+                      '(string string string))
```

Comment:
This is fine for now.  For performance and compactness, ultimately what I'd like to do is factor out each complex (variable-length, like strings) value type into its own helper relation, and replace their references with ids pointing into those respective tables.

So we'd have individual relations for the components of `concept`, such as `concept-curie`, `concept-property-name`, and `concept-property-value`, and then define an underlying low-level `concept-representation` relation as a triple of ids.  The primary and index tables of `concept-representation` would then be significantly more compact and faster to reference.  We can then define a high-level `concept` relation on top of all this, that exposes the string-based interface, and does the appropriate joins for us.  When dbKanren is more mature, it will be able to eliminate any overhead from this abstraction (using functional dependencies), and automatically make use of the low-level relations appropriately in response to constraints on arguments to calls to the high-level relation.

We'd do likewise for `predicate` and `predicateprop`.